rfcs/0007-handling-corruption.md

```diff
@@ -217,108 +209,184 @@ durability and availability systems on top of SlateDB.
+operation to continue, SlateDB should endeavor to return the most recent 
+valid data it can.
+
+- If the SST Bloom Filter indicates a value exists in the SST, yet the key was
+  not found, and corruption was detected during iteration, the `get` caller
+  should receive a Corruption error.
+- If the most recent value is in a non-corrupt block, the `get` call should
+  return successfully
+- If the SST index is corrupt and `get` has not yet found the KV in an earlier
+  SST, the call should receive a Corruption error. (This effectively prevents
+  reads in SSTs older than the first corrupt SST)
+- MORE?
+
+#### Corruption Inspection and Repair
+To support the repair of SSTs, I propose introducing two new functions:
+`sst_inspect()` and `sst_repair()`. These functions will provide developers with
+tools to inspect the current contents of a specific SST, and repair a specific
+SST that has been identified as corrupt.
+
+To support the repair of the WAL, I propose introducing two additional functions:
+`wal_inspect()` and `wal_repair()`. These functions will allow developers to
+inspect the current contents of the WAL, including corrupted entries, and to
+repair the WAL to the best of its ability through entity recovery or truncation
+of the WAL.
+
+The inspection functions are designed to provide verbose information about the
+WAL or SST. This includes reporting the contents of the files and any
+corruption found. The goal is to provide operators with tools to inspect the 
+file contents, any corruption, and the surrounding data without needing intimate
+knowledge of the file format.
+
+The inspect and repair functions should share the following characteristics:
+- Both WAL and SST variants should accept raw file input, such that procurement
+  of the file via SST ID (either WAL or compacted) or out of band file procurement
+  is possible. This allows operators or developers to download files out of band
+  and inspect or repair the files using the provided functions
+- Repair functions should include a "dry run" option which only reports repairs 
+without actually perform the repair.
+
+We should introduce a new function called `repair()`, which encapsulates both
+`sst_repair` and `wal_repair` into a single function. This function scans the WAL
+and all SSTs in the database to inspect them for corruption and optionally
+repairs them if the "dry run" option is not provided.
+
+Before performing any repair, the `repair()` function will make a copy of the
+original file and store it in a `corrupted/` path on the object store. This
+provides operators who run the `repair()` function in haste with the option to
+inspect the corrupt data file after the database has been restored to an
+operational state.
+
+#### For Convenience
+For convenience:
+- We could optionally introduce a new database option called
+  `DbOptions.wal_auto_recovery`, similar in naming to `DbOptions.wal_enabled`.
```

Comment:
WDYT about `DbOptions.sst_auto_recovery`, which would auto recover both WAL and compacted SSTs? I suspect users that want WAL auto recovery would also want compacted SST auto recovery.

Reply:
I just realized that `DbOptions.wal_auto_recovery` does not make sense.... a `DBOptions.on_corruption` function that does not return an error, causes SlateDB to skip corrupt WAL entries, which is exactly what `wal_repair()` will do. Hence, if you want WAL auto recovery, just define an `DBOptions.on_corruption` callback which does not return an error.

`DbOptions.sst_auto_recovery` doesn't make sense unless we are verifying the integrity of the entire database during `open_with_opts()`. Just like WAL recovery, recovering SST's is essentially skipping corrupted entries. You can get the same result by defining a `DBOptions.on_corruption` callback which does not return an error.

Reply:
Fewer configs! 🎉 I love it.